### PR TITLE
Throw a better exception when a property was not discovered, but could be a primitive property

### LINKED ISSUE #4484: Property mapping validation provides unhelpful message for primitive type that cannot be mapped
If the entity has a property of a type that cannot be mapped by the provider and is not a primitive type, then the exception message is:
`The property '{property}' on entity type '{entityType}' has CLR type which is not supported by current provider and it has not been configured to use any supported column type.`
However, if the property is a primitive type, such as uint for SQL Server, then the message is:
`The property '{property}' on entity type '{entityType}' has not been added to the model or ignored.`

It seems to me that we should not be treating primitive types differently in this case.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Throw a better exception when a property of interface type was not discovered

Resolves #4484, #4627

### CODE REVIEW

**Comment by AndriySvyryd** [comment]
@rowanmiller for exception message review

**Comment by ajcvickers** [comment]
:shipit: I like the new messages.

**Comment by rowanmiller** [comment]
Can you summarize when these get thrown? I think we can improve the "has not been added to the model" wording as we add all properties to the model by default. Perhaps it's more a case of "could not be added to the model". I can work on the wording if you just let me know the situations in which each message is thrown :smile:

**Comment by AndriySvyryd** [comment]
@rowanmiller Currently we only discover and add primitive properties that can be mapped. If such property is present on the CLR type or if it was added explicitly, but without specifying the provider type `PropertyNotMapped` is thrown.
`NavigationNotAdded` is thrown for properties that could be navigations, but weren't discovered, probably because there's an ambiguity matching the inverse, convention logging would show more info.
`InterfacePropertyNotAdded` is for interface type properties, we never discover them.
`PropertyNotAdded` is for everything else that still could be a property, so mostly value types that aren't considered primitive in Core or `object`.

File: src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx [hunk 1]
```diff
@@ -466,13 +466,16 @@
     <value>The entity type '{entityType}' cannot be removed because '{derivedEntityType}' is derived from it. All derived entity types must be removed or redefined before the entity type can be removed.</value>
   </data>
   <data name="NavigationNotAdded" xml:space="preserve">
-    <value>The navigation '{navigation}' on entity type '{entityType}' has not been added to the model, or ignored, or target entityType ignored.</value>
+    <value>The navigation property '{navigation}' of type '{propertyType}' on entity type '{entityType}' has not been added to the model. Either manually configure a relationship using this property, ignore the property or ignore the target entity type.</value>
   </data>
```
Comment (rowanmiller) [inline src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx]: I think the implementation details of what it means to be "added to the model" are leaking out here. Conceptually (for the app developer) all properties (with a public getter/setter are included in the model) but EF can't work out what to do with this one.

Here is my suggestion:

> Unable to determine the relationship represented by navigation property '{entityType}.{navigation} of type '{propertyType}'. Either manually configure the relationship, or ignore this property from the model.

File: src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx [hunk 2]
```diff
@@ -466,13 +466,16 @@
...
   <data name="PropertyNotAdded" xml:space="preserve">
-    <value>The property '{property}' on entity type '{entityType}' has not been added to the model or ignored.</value>
+    <value>The property '{property}' on entity type '{entityType}' has not been added to the model, because it is of type '{propertyType}' which is not recognized as a supported primitive type or a valid entity type. Either explicitly map this property or ignore it.</value>
   </data>
```
Comment (rowanmiller) [inline src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx]: > The property '{entityType}.{property}' could not be mapped, because it is of type '{propertyType}' which is not a supported primitive type or a valid entity type. Either explicitly map this property, or ignore it.

File: src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx [hunk 3]
```diff
@@ -466,13 +466,16 @@
...
   <data name="PropertyNotMapped" xml:space="preserve">
-    <value>The property '{property}' on entity type '{entityType}' has CLR type which is not supported by current provider and it has not been configured to use any supported column type.</value>
+    <value>The property '{property}' on entity type '{entityType}' is of type '{propertyType}' which is not supported by current provider. Either change the property CLR type or manually configure the provider type for it.</value>
+  </data>
```
Comment (rowanmiller) [inline src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx]: > The property '{entityType}.{property}' is of type '{propertyType}' which is not supported by current database provider. Either change the property CLR type or manually configure the database type for it.

File: src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx [hunk 4]
```diff
@@ -466,13 +466,16 @@
...
+  <data name="InterfacePropertyNotAdded" xml:space="preserve">
+    <value>The property '{navigation}' of type '{propertyType}' on entity type '{entityType}' has not been added to the model. If it is a navigation property manually configure a relationship using this property by casting it to a mapped entity type otherwise ignore the property.</value>
   </data>
```
Comment (rowanmiller) [inline src/Microsoft.EntityFrameworkCore/Properties/CoreStrings.resx]: > The property '{entityType}.{navigation}' is of an interface type ('{propertyType}'). If it is a navigation property manually configure the relationship for this property by casting it to a mapped entity type, otherwise ignore the property from the model.

**Comment by rowanmiller** [comment]
Suggested some updated exception messages that I think match better with what is conceptually happening for the app developer rather than what is happening in our stack. Let me know if you have pushback.

**Comment by divega** [comment]
:+1: I like the suggested messages and the conceptual approach.

**Comment by smitpatel** [comment]
:+1:
